pkg/watch/health.go

```diff
@@ -0,0 +1,239 @@
+	lastCheck  time.Time          // time of last health check
+	lastStatus health.HealthState // status of last health check
+
+	logger *logging.Logger
+}
+
+// MonitorPodHealth is meant to be a long running go routine.
+// MonitorPodHealth reads from a consul store to determine which
+// services should be running on the host. MonitorPodHealth
+// runs a CheckHealth routine to monitor the health of each
+// service and kills routines for services that should no
+// longer be running.
+func MonitorPodHealth(config *preparer.PreparerConfig, logger *logging.Logger, shutdownCh chan struct{}) {
+	var store kp.Store
+
+	consul := config.ConsulAddress
+	node := config.NodeName
+	pods := []PodWatch{}
+	authtoken, err := preparer.LoadConsulToken(config.ConsulTokenPath)
+	if err != nil {
+		logger.WithField("inner_err", err).Warningln("Could not load consul token")
+	}
+
+	store = kp.NewConsulStore(kp.Options{
+		Address: consul,
+		HTTPS:   false,
+		Token:   authtoken,
+		Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+	})
+	pods = updateHealthMonitors(store, pods, node, logger)
+	for {
+		select {
+		case <-time.After(POLL_KV_FOR_PODS):
+			// check if pods have been added or removed
+			// starts monitor routine for new pods
+			// kills monitor routine for removed pods
+			pods = updateHealthMonitors(store, pods, node, logger)
+		case <-shutdownCh:
+			return
+		}
+	}
+}
+
+// Monitor Health is a go routine that runs as long as the
+// service it is monitoring. Every HEALTHCHECK_INTERVAL it
+// performs a health check and writes that information to
+// consul
+func (p *PodWatch) MonitorHealth(node string, store kp.Store, shutdownCh chan bool) {
+	for {
+		select {
+		case <-time.After(HEALTHCHECK_INTERVAL):
+			p.checkHealth(node, p.manifest.StatusPort, store)
+		case <-shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *PodWatch) checkHealth(node string, port int, store kp.Store) {
+	resp := kp.HttpStatusCheck(node, port)
+	health, err := resultFromCheck(resp)
+	if err != nil {
+		return
+	}
+	health.ID = p.manifest.Id
+	health.Node = node
+	health.Service = p.manifest.Id
+
+	if p.updateNeeded(health, TTL) {
+		p.lastCheck, err = writeToConsul(health, store)
+		p.lastStatus = health.Status
+		if err != nil {
+			p.logger.WithField("err", err).Warningln("failed to write health data to consul")
+		}
+	}
+}
+
+func updateHealthMonitors(store kp.Store, watchedPods []PodWatch, node string, logger *logging.Logger) []PodWatch {
+	path := kp.RealityPath(node)
+	reality, _, err := store.ListPods(path)
+	if err != nil {
+		logger.WithField("inner_err", err).Warningln("failed to get pods from reality store")
+	}
+
+	return updatePods(watchedPods, reality, logger, store, node)
+}
+
+func resultFromCheck(resp *http.Response) (health.Result, error) {
+	var err error
+	res := health.Result{}
+	if resp == nil {
+		res.Status = health.Critical
+		return res, nil
+	}
+	if resp.StatusCode == 200 {
+		res.Status = health.Passing
```

Comment:
I think we do care. It's an easy feature to provide and someone will ask for it.